src/reflect/scala/reflect/internal/tpe/TypeComparers.scala

```diff
@@ -381,18 +381,23 @@ trait TypeComparers {
   // @assume tp1.isHigherKinded || tp2.isHigherKinded
   def isHKSubType(tp1: Type, tp2: Type, depth: Depth): Boolean = {
 
+    def hkSubVariance(tparams1: List[Symbol], tparams2: List[Symbol]) =
+      (tparams1 corresponds tparams2)(methodHigherOrderTypeParamsSubVariance)
+
     def isSubHKTypeVar(tp1: Type, tp2: Type) = (tp1, tp2) match {
       case (tv1: TypeVar, tv2: TypeVar) =>
         devWarning(sm"Unexpected code path: testing two type variables for subtype relation: $tv1 <:< $tv2")
         tv1 eq tv2
       case (_, tv2: TypeVar) =>
         val ntp1 = tp1.normalize
-        (tv2.params corresponds ntp1.typeParams)(methodHigherOrderTypeParamsSubVariance) &&
-        { tv2.addLoBound(ntp1); true }
+        val kindsMatch = (ntp1.typeSymbol eq AnyClass) || hkSubVariance(tv2.params, ntp1.typeParams)
+        if (kindsMatch) tv2.addLoBound(ntp1)
+        kindsMatch
```

Comment:
I find constructs of the sort `foo && { bar(); true }` very confusing. Hope this makes the intent clearer.

Reply:
I agree with the change. Mixing boolean operator's behaviours as boolean functions and control-flow statements can be confusing.